bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/infoproviders/AddonsInfoProvider.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.addon.infoproviders;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.addon.AddonDiscoveryMethod;
+import org.openhab.core.addon.AddonInfo;
+import org.openhab.core.addon.AddonInfoListReader;
+import org.openhab.core.addon.AddonInfoProvider;
+import org.openhab.core.addon.AddonMatchProperty;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.thoughtworks.xstream.XStreamException;
+import com.thoughtworks.xstream.converters.ConversionException;
+
+/**
+ * The {@link AddonsInfoProvider} reads all {@code userdata/addons/*.xml} files, each of which
+ * should contain a list of addon.xml elements, and convert their combined contents into a list
+ * of {@link AddonInfo} objects can be accessed via the {@link AddonInfoProvider} interface.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = AddonInfoProvider.class, name = AddonsInfoProvider.SERVICE_NAME)
+public class AddonsInfoProvider implements AddonInfoProvider {
+
+    public static final String SERVICE_NAME = "addons-info-provider";
+
+    private static final boolean TEST_ADDON_DEVELOPER_REGEX_SYNTAX = true;
+
+    private final Logger logger = LoggerFactory.getLogger(AddonsInfoProvider.class);
+    private final String folder = OpenHAB.getUserDataFolder() + File.separator + "addons";
+    private final Set<AddonInfo> addonInfos = new HashSet<>();
+
+    @Activate
+    public AddonsInfoProvider() {
+        initialize();
```

Comment:
How long could that take? It might not the the ideal place to do this in the constructor as it slows down the system startup. Could it be scheduled in a separate thread?

Reply:
> Could it be scheduled in a separate thread?

Thinking about this further: I think this constructor must block until the AddonInfos have been loaded from the .xml disk file. It is essential that this ('source') component is fully initialized before the scanner ('consumer') components would be loaded and start searching for suggested addons. (Or we could try some complex cross synchronization across the 'source' and 'consumer' components but either way, it would take the same amount of time to start).